include/gsl/gsl-lite.hpp

```diff
@@ -2220,6 +2226,72 @@ not_null<U> make_not_null( not_null<U> const & u )
 }
 #endif // gsl_HAVE( RVALUE_REFERENCE )
 
+namespace detail
+{
+
+// Generic case handling anything that is not a not_null
+template< class CVReference, class T /* = typename std20::remove_cvref<CVReference>::type */>
+struct as_nullable_helper
+{
+    typedef T nullable_type;
+
+    static nullable_type const & call(nullable_type const & p)
+    {
+        return p;
+    }
+
+#if gsl_HAVE( RVALUE_REFERENCE )
+    static nullable_type && call(nullable_type && p)
+    {
+        return std::move(p);
+    }
+#endif
+};
+
+// Specific case handling not_null
+template< class CVReference, class T >
+struct as_nullable_helper< CVReference, not_null<T> >
+{
+    typedef T nullable_type;
+
+    static nullable_type const & call(not_null<nullable_type> const & p)
+    {
+        return p.data_.ptr_;
+    }
+
+#if gsl_HAVE( RVALUE_REFERENCE )
+    static nullable_type && call(not_null<nullable_type> && p)
```

Comment:
This function should return `nullable_type`, not `nullable_type &&`, otherwise `as_nullable(std::move(nnp))` can be abused to get direct access to the underlying pointer.